README
NAME
    Module::Install::RTx - RT extension installer

SYNOPSIS
    In the Makefile.PL of the "RT-Extension-Example" module:

        use inc::Module::Install;
        RTx 'RT-Extension-Example';

        requires_rt '4.2.0';
        rt_too_new  '4.4.0';

        WriteAll();

DESCRIPTION
    This Module::Install extension implements several functions for
    installing RT extensions:

  RTx '*extension name*'
    This function arranges for the following directories to be installed, if
    they exist (assuming "RTx('RT-Extension-Example')"):

        ./bin       => $RT::LocalPluginPath/RT-Extension-Example/bin
        ./etc       => $RT::LocalPluginPath/RT-Extension-Example/etc
        ./html      => $RT::LocalPluginPath/RT-Extension-Example/html
        ./lib       => $RT::LocalPluginPath/RT-Extension-Example/lib
        ./po        => $RT::LocalPluginPath/RT-Extension-Example/po
        ./sbin      => $RT::LocalPluginPath/RT-Extension-Example/sbin
        ./static    => $RT::LocalPluginPath/RT-Extension-Example/static
        ./var       => $RT::LocalPluginPath/RT-Extension-Example/var
        ./patches   => If this directory/sub directory(s) contain one or more
                       p1 unified diffs they will be applied to the RT
                       installation. Invoked in your plugin as make patch

    Accepts an optional argument hashref after the extension name with two
    possible keys

  deprecated_rt
        If set to a true value, skips the enforced RT-4.0.0 minimum version check

        You should set a perl_version if using this option and requires_rt(), because requires_rt
        only handles figuring our what perl you need if you're on RT 4.0.0 or higher.

  no_readme_generation
        If set to a true value, will not call readme_from on the extension's primary perl module.

  requires_rt *version*
    Takes one argument, a valid RT version. If an attempt is made to install
    on an RT than that version, it will die before Makefile creation.

  requires_rt_plugin *RT::Extension::Example* [, *version*]
    Ensures that the given RT extension (and optional version) is installed
    in the target RT instance; "requires" cannot be used because RT
    extensions are not in @INC.

  rt_too_new *version* [, *message*]
    Takes one argument, a valid RT version, and prevents this module from
    being installed on any version of RT equal to or newer than that. Useful
    if a particular release of an extension only works on 4.0.x but not
    4.2.x.

    Takes an optional second argument which allows you to specify a custom
    error message. This message is passed to sprintf with two string
    arguments, the current RT version and the version you specify.

CAVEATS
    *   Us the full name when calling RTx method in Makefile.PL; while
        "RTx('Foo')" was once supported, it is no longer.

ENVIRONMENT
    RTHOME
        Path to the RT installation that contains a valid lib/RT.pm.

EXAMPLES
    To install an extension which makes use of this installer:

        perl Makefile.PL RTHOME=/opt/rt4

    This will install all subdirs into the $RT::LocalPluginPath dir as
    configured in RT::Generated.

    To install an extension into the (vendor) plugin path:

        perl Makefile.PL RTHOME=/opt/rt4 INSTALLDIRS=vendor

    This will install all subdirs into the $RT::PluginPath which is
    specifically meant for plugins that are installed through other
    packaging utils like APT or RPM.

SEE ALSO
    Module::Install

    <http://www.bestpractical.com/rt/>

AUTHORS
    Best Practical Solutions

    (Originally) Audrey Tang <cpan@audreyt.org>

COPYRIGHT
    Copyright 2003, 2004, 2007 by Audrey Tang <cpan@audreyt.org>. Copyright
    2008-2014 Best Practical Solutions

    This software is released under the MIT license cited below.

  The "MIT" License
    Permission is hereby granted, free of charge, to any person obtaining a
    copy of this software and associated documentation files (the
    "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be included
    in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
    OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



Changes
0.40
 - Add . to @INC for "make initdb" to help Perl 5.26+ load etc/initialdata

0.39 2017-02-20
- Support `remove_files => 1` parameter to force unlinking of files
  provided by previous versions

0.38 2016-07-14
 - Support `INSTALLDIRS=vendor`
 - Properly honor `$DESTDIR` at `make install` time

0.37 2014-12-24
 - Add the Module::Install::ReadmeFromPod dependency

0.36 2014-09-02
 - Add deprecated_rt and no_readme_generation flags to RTx()
 - Alter the right structures when requires_rt_plugin is called twice

0.35 2014-07-28
 - No changes since 0.34_05

0.34_05 2014-07-14
 - Apply consistent transformations to RTHOME as we do when prompting
 - Check Debian request-tracker4 install location by default as well

0.34_04 2014-06-26
 - Automatically generate a README

0.34_03 2014-06-19
 - Stop runtime code from clearing out inc/ if run as an author

0.34_02 2014-06-19
 - Fix requires_rt_plugin
 - Standardize error messages for rt_too_new and requires_rt

0.34_01 2014-06-19
 - Remove RT 3.8-specific code
 - Update documentation to reflect 3.8-era Plugin world order
 - Add requires_rt_plugin helper
 - Remove WITH_SUBDIRS support
 - Remove non-working PREFIX support
 - Add RT version information to META.yml
 - Default minimum perl version based on minimum RT version
 - Tidy Changes file

0.34 2014-05-20
 - Backcompat workaround so you can ship static/css and static/js and on
   4.0 we copy to html/NoAuth/{css,js}.  Isn't a full win for images which
   needed a dhandler hack under 4.0.

0.33 2014-03-17
 - Better error message when requires_rt dies
 - Release a pent up set of changes now that they've been tested

0.32_02 2014-02-10
 - Avoid some false positives that would offer a make upgrade-database
   which cannot work.

0.32_01 2014-01-10
 - Now that we aren't packaging extensions against 3.6, we can use
   cmp_versions imported from RT core rather than copy/paste.
 - new rt_too_new which allows you to bail out before being installed on a
   version of RT which is too new for you to function normally on
   (corollary to requires_rt).
 - requires_rt now dies if your RT is too old rather than just warning.

0.32 2013-09-06
 - new make upgrade-database command; this feature needs RT 4.2 or
   4.0.14+ for maximum benefit, and is used in the RTIR 3.0 release
   series

0.31 2013-04-08
 - Use the DatabaseAdmin configuration if supplied
 - Install /static/ on future 4.2 versions of RT

0.30 2012-11-06
 - remove a lot of old unused code
 - add support for a 4.1 feature that records database changes made by the
   extension

0.29_02 2012-08-16
 - further removal of make factory, this time 'make dropdb'

0.29_01 2012-08-16
 - Remove support for 'make factory'
 - Add support for a new 4.1 feature that records database changes made by
   the extension

0.29 2011-05-20
 - add requires_rt to handle requiring an RT version since MakeMaker and
   Module::Install::AutoInstall can't deal with RT version numbers.

0.28 2011-03-15
 - search /opt/rt4 for RT.pm

0.27 2010-12-16
 - Include RT's lib/ directory in INC so that requires() on RT::VERSION
   work properly

0.26 2010-12-14
 - Search RTHOME before searching @INC This works around distribution
   packages lurking in @INC

0.25 2009-10-08
 - Make prompt for lib path clearer [rt.cpan.org #50328
 - trim RT.pm from lib paths entered by users

0.24 2008-09-15
 - updated docs
 - die when target is RT 3.8.0
 - install libs only into one path, don't double-install them on RT 3.8.x
 - delete VERSION from POD, we just always forget to update it

0.23 2008-05-09
 - RT now respects the --datafile flag to rt-setup-database when used with
   --action acl or --action schema. Don't pass the initialdata file when
   we're doing schema changes

0.22 2008-03-25
 - Add local lib path to the @INC when we call any script that may load RT
   config.

0.21 2007-12-07
 - LICENSING CHANGE: This compilation and all individual files in it are
   now under the permissive "MIT" license.  See the COPYRIGHT section in
   README for the new terms.
 - Cleanup on POD, changelog, README, etc.
 - Bump dependency of Module::Install::Admin to 0.40, so we can make use of
   all_from() and drop the parentheses in Makefile.PL commands.

0.20 2007-12-05
 - Support for new RT 3.7 "RT Plugin" mechanism.
 - No longer tries to write to the global local perl man and site
   directories (packlists and manpages), instead installing them within the
   RT hierarchy.

0.11 2005-02-26
 - make initdb now prompts for dba password.

0.10 2004-09-09
 - "make initialize-database" added as an alias to "make initdb", for
   better consistency with core RT installation. Reported by: Sika.

0.09 2004-09-09
 - Moved under SVK management; remove all keyword expansion lines.
 - Corrected the use of $ENV{RTHOME} and documented it.

0.08 2004-05-31
 - "make initdb" now moved to M::I::RTx::Factory, so the person running
   "perl Makefile.PL" needs not have permission to read RT_SiteConfig.pm.
 - M::I::RTx::Factory now survives pod-stripping.
 - Jesse reports that RTx('RT-Foo-Bar') breaks.
 - Improve table detection in Pg.
 - Better handling failed require() on _Overlay classes in the factory.

0.07 2004-05-13
 - Need to use RT::LoadConfig instead of requiring the config files
   directly. Reported by: Jesse Vincent

0.06 2004-05-10
 - Support "make factory", "make initdb", "make dropdb" for extensions that
   has its own initialdata and schema.

0.05 2004-02-01
 - Now installs libs to $RT::LocalPath/lib instead of perl's sitelib.
 - Now properly skips lib installation when WITH_SUBDIRS does not include
   lib.
 - Allow overriding prefix with $ENV{PREFIX}

0.04 2004-01-10
 - Update copyright years.
 - Make var/ directory writable
 - Some people insist on doing "make initialize-database" before "make
   install".
 - Add etc/initialdata insertion.
 - Take care of "inplace" layout by also probing in lib/RT.pm.
 - Now takes WITH_SUBDIRS to restrict the subdirectories to install.
 - Some POD cleanups.

0.01 2003-12-15

lib/Module/Install/RTx/Factory.pm
package Module::Install::RTx::Factory;
use Module::Install::Base; @ISA = qw(Module::Install::Base);

use strict;
use File::Basename ();

sub RTxInitDB {
    my ($self, $action, $name, $version) = @_;

    unshift @INC, substr(delete($INC{'RT.pm'}), 0, -5) if $INC{'RT.pm'};

    require RT;
    unshift @INC, "$RT::LocalPath/lib" if $RT::LocalPath;

    $RT::SbinPath ||= $RT::LocalPath;
    $RT::SbinPath =~ s/local$/sbin/;

    foreach my $file ($RT::CORE_CONFIG_FILE, $RT::SITE_CONFIG_FILE) {
        next if !-e $file or -r $file;
        die "No permission to read $file\n-- please re-run $0 with suitable privileges.\n";
    }

    RT::LoadConfig();

    require RT::System;

    my $lib_path = File::Basename::dirname($INC{'RT.pm'});
    my @args = ("-Ilib");
    push @args, "-I$RT::LocalPath/lib" if $RT::LocalPath;
    push @args, (
        "-I$lib_path",
        "$RT::SbinPath/rt-setup-database",
        "--action"      => $action,
        "--datadir"     => "etc",
        (($action eq 'insert') ? ("--datafile"    => "etc/initialdata") : ()),
        "--dba"         => $RT::DatabaseAdmin || $RT::DatabaseUser,
        "--prompt-for-dba-password" => '',
        (RT::System->can('AddUpgradeHistory') ? ("--package" => $name, "--ext-version" => $version) : ()),
    );
    if ($action eq 'upgrade' and
        not RT::System->can('AddUpgradeHistory')) {
        push @args, "--package" => $name;
    }

    print "$^X @args\n";
    (system($^X, @args) == 0) or die "...returned with error: $?\n";
}

1;


lib/Module/Install/RTx/Runtime.pm
package Module::Install::RTx::Runtime;

use base 'Exporter';
our @EXPORT = qw/RTxDatabase RTxConfig RTxPlugin/;

use strict;
use File::Basename ();
use File::Copy;

sub _rt_runtime_load {
    require RT;

    eval { RT::LoadConfig(); };
    if (my $err = $@) {
        die $err unless $err =~ /^RT couldn't load RT config file/m;
        my $warn = <<EOT;
This usually means that your current user cannot read the file.  You
will likely need to run this installation step as root, or some user
with more permissions.
EOT
        $err =~ s/This usually means.*/$warn/s;
        die $err;
    }
}

sub RTxDatabase {
    my ($action, $name, $version) = @_;

    _rt_runtime_load();

    require RT::System;
    my $has_upgrade = RT::System->can('AddUpgradeHistory');

    my $lib_path = File::Basename::dirname($INC{'RT.pm'});
    my @args = (
        "-Ilib",
        "-I$RT::LocalLibPath",
        "-I$lib_path",
        "$RT::SbinPath/rt-setup-database",
        "--action"      => $action,
        ($action eq 'upgrade' ? () : ("--datadir"     => "etc")),
        (($action eq 'insert') ? ("--datafile"    => "etc/initialdata") : ()),
        "--dba"         => $RT::DatabaseAdmin || $RT::DatabaseUser,
        "--prompt-for-dba-password" => '',
        ($has_upgrade ? ("--package" => $name, "--ext-version" => $version) : ()),
    );
    # If we're upgrading against an RT which isn't at least 4.2 (has
    # AddUpgradeHistory) then pass --package.  Upgrades against later RT
    # releases will pick up --package from AddUpgradeHistory.
    if ($action eq 'upgrade' and not $has_upgrade) {
        push @args, "--package" => $name;
    }

    print "$^X @args\n";
    (system($^X, @args) == 0) or die "...returned with error: $?\n";
}

sub RTxConfig {
    my $name = shift;

    _rt_runtime_load();

    my $src = join('/', $RT::LocalPluginPath, $name, 'etc', $name);
    my $dst = join('/', $RT::EtcPath, 'RT_SiteConfig.d', $name);

    $src .= ".pm";
    $dst .= ".pm";

    copy($src, $dst) or die "Install of configuration failed: $!\n"
                            . "You did run make install first??";

}

sub RTxPlugin {
    my ($name) = @_;

    _rt_runtime_load();
    require YAML::Tiny;
    my $data = YAML::Tiny::LoadFile('META.yml');
    my $name = $data->{name};

    my @enabled = RT->Config->Get('Plugins');
    for my $required (@{$data->{x_requires_rt_plugins} || []}) {
        next if grep {$required eq $_} @enabled;

        warn <<"EOT";

**** Warning: $name requires that the $required plugin be installed and
              enabled; it is not currently in \@Plugins.

EOT
    }
}

1;


lib/Module/Install/RTx.pm
package Module::Install::RTx;

use 5.008;
use strict;
use warnings;
no warnings 'once';

use Module::Install::Base;
use base 'Module::Install::Base';
our $VERSION = '0.39';

use FindBin;
use File::Glob     ();
use File::Basename ();

my @DIRS = qw(etc lib html static bin sbin po var);
my @INDEX_DIRS = qw(lib bin sbin);

sub RTx {
    my ( $self, $name, $extra_args ) = @_;
    $extra_args ||= {};

    # Set up names
    my $fname = $name;
    $fname =~ s!-!/!g;

    $self->name( $name )
        unless $self->name;
    $self->all_from( "lib/$fname.pm" )
        unless $self->version;
    $self->abstract("$name Extension")
        unless $self->abstract;
    unless ( $extra_args->{no_readme_generation} ) {
        $self->readme_from( "lib/$fname.pm",
                            { options => [ quotes => "none" ] } );
    }
    $self->add_metadata("x_module_install_rtx_version", $VERSION );

    my $installdirs = $ENV{INSTALLDIRS};
    for ( @ARGV ) {
        if ( /INSTALLDIRS=(.*)/ ) {
            $installdirs = $1;
        }
    }

    # Try to find RT.pm
    my @prefixes = qw( /opt /usr/local /home /usr /sw /usr/share/request-tracker4);
    $ENV{RTHOME} =~ s{/RT\.pm$}{} if defined $ENV{RTHOME};
    $ENV{RTHOME} =~ s{/lib/?$}{}  if defined $ENV{RTHOME};
    my @try = $ENV{RTHOME} ? ($ENV{RTHOME}, "$ENV{RTHOME}/lib") : ();
    while (1) {
        my @look = @INC;
        unshift @look, grep {defined and -d $_} @try;
        push @look, grep {defined and -d $_}
            map { ( "$_/rt4/lib", "$_/lib/rt4", "$_/lib" ) } @prefixes;
        last if eval {local @INC = @look; require RT; $RT::LocalLibPath};

        warn
            "Cannot find the location of RT.pm that defines \$RT::LocalPath in: @look\n";
        my $given = $self->prompt("Path to directory containing your RT.pm:") or exit;
        $given =~ s{/RT\.pm$}{};
        $given =~ s{/lib/?$}{};
        @try = ($given, "$given/lib");
    }

    print "Using RT configuration from $INC{'RT.pm'}:\n";

    my $local_lib_path = $RT::LocalLibPath;
    unshift @INC, $local_lib_path;
    my $lib_path = File::Basename::dirname( $INC{'RT.pm'} );
    unshift @INC, $lib_path;

    # Set a baseline minimum version
    unless ( $extra_args->{deprecated_rt} ) {
        $self->requires_rt('4.0.0');
    }

    # Installation locations
    my %path;
    my $plugin_path;
    if ( $installdirs && $installdirs eq 'vendor' ) {
        $plugin_path = $RT::PluginPath;
    } else {
        $plugin_path = $RT::LocalPluginPath;
    }
    $path{$_} = $plugin_path . "/$name/$_"
        foreach @DIRS;

    # Copy RT 4.2.0 static files into NoAuth; insufficient for
    # images, but good enough for css and js.
    $path{static} = "$path{html}/NoAuth/"
        unless $RT::StaticPath;

    # Delete the ones we don't need
    delete $path{$_} for grep {not -d "$FindBin::Bin/$_"} keys %path;

    my %index = map { $_ => 1 } @INDEX_DIRS;
    $self->no_index( directory => $_ ) foreach grep !$index{$_}, @DIRS;

    my $args = join ', ', map "q($_)", map { ($_, "\$(DESTDIR)$path{$_}") }
        sort keys %path;

    printf "%-10s => %s\n", $_, $path{$_} for sort keys %path;

    if ( my @dirs = map { ( -D => $_ ) } grep $path{$_}, qw(bin html sbin etc) ) {
        my @po = map { ( -o => $_ ) }
            grep -f,
            File::Glob::bsd_glob("po/*.po");
        $self->postamble(<< ".") if @po;
lexicons ::
\t\$(NOECHO) \$(PERL) -MLocale::Maketext::Extract::Run=xgettext -e \"xgettext(qw(@dirs @po))\"
.
    }

    my $remove_files;
    if( $extra_args->{'remove_files'} ){
        $self->include('Module::Install::RTx::Remove');
        our @remove_files;
        eval { require "etc/upgrade/remove_files" }
          or print "No remove file located, no files to remove\n";
        $remove_files = join ",", map {"q(\$(DESTDIR)$plugin_path/$name/$_)"} @remove_files;
    }

    $self->include('Module::Install::RTx::Runtime') if $self->admin;
    $self->include_deps( 'YAML::Tiny', 0 ) if $self->admin;
    my $postamble = << ".";
install ::
\t\$(NOECHO) \$(PERL) -Ilib -I"$local_lib_path" -I"$lib_path" -Iinc -MModule::Install::RTx::Runtime -e"RTxPlugin()"
.

    if( $remove_files ){
        $postamble .= << ".";
\t\$(NOECHO) \$(PERL) -MModule::Install::RTx::Remove -e \"RTxRemove([$remove_files])\"
.
    }

    $postamble .= << ".";
\t\$(NOECHO) \$(PERL) -MExtUtils::Install -e \"install({$args})\"
.

    if ( $path{var} and -d $RT::MasonDataDir ) {
        my ( $uid, $gid ) = ( stat($RT::MasonDataDir) )[ 4, 5 ];
        $postamble .= << ".";
\t\$(NOECHO) chown -R $uid:$gid $path{var}
.
    }

    my %has_etc;
    if ( File::Glob::bsd_glob("$FindBin::Bin/etc/*.pm") ) {
        $has_etc{config}++;
    }
    if ( File::Glob::bsd_glob("$FindBin::Bin/etc/schema.*") ) {
        $has_etc{schema}++;
    }
    if ( File::Glob::bsd_glob("$FindBin::Bin/etc/acl.*") ) {
        $has_etc{acl}++;
    }
    if ( -e 'etc/initialdata' ) { $has_etc{initialdata}++; }
    if ( grep { /\d+\.\d+\.\d+.*$/ } glob('etc/upgrade/*.*.*') ) {
        $has_etc{upgrade}++;
    }

    $self->postamble("$postamble\n");
    if ( $path{lib} ) {
        $self->makemaker_args( INSTALLSITELIB => $path{'lib'} );
        $self->makemaker_args( INSTALLARCHLIB => $path{'lib'} );
        $self->makemaker_args( INSTALLVENDORLIB => $path{'lib'} )
    } else {
        $self->makemaker_args( PM => { "" => "" }, );
    }

    $self->makemaker_args( INSTALLSITEMAN1DIR => "$RT::LocalPath/man/man1" );
    $self->makemaker_args( INSTALLSITEMAN3DIR => "$RT::LocalPath/man/man3" );
    $self->makemaker_args( INSTALLSITEARCH => "$RT::LocalPath/man" );

    # INSTALLDIRS=vendor should install manpages into /usr/share/man.
    # That is the default path in most distributions. Need input from
    # Redhat, Centos etc.
    $self->makemaker_args( INSTALLVENDORMAN1DIR => "/usr/share/man/man1" );
    $self->makemaker_args( INSTALLVENDORMAN3DIR => "/usr/share/man/man3" );
    $self->makemaker_args( INSTALLVENDORARCH => "/usr/share/man" );

    if (%has_etc) {
        if ($has_etc{config}) {
            print "For first-time installation, type 'make init-config'.\n";
            my $initconfig = qq|\t\$(NOECHO) \$(PERL) -Ilib -I"$local_lib_path" -I"$lib_path" -Iinc -MModule::Install::RTx::Runtime -e"RTxConfig(qw(\$(DISTNAME)))"\n|;
            $self->postamble("init-config ::\n$initconfig\n");
        }

        print "For first-time installation, type 'make initdb'.\n";
        my $initdb = '';
        $initdb .= <<"." if $has_etc{schema};
\t\$(NOECHO) \$(PERL) -Ilib -I"$local_lib_path" -I"$lib_path" -Iinc -MModule::Install::RTx::Runtime -e"RTxDatabase(qw(schema \$(NAME) \$(VERSION)))"
.
        $initdb .= <<"." if $has_etc{acl};
\t\$(NOECHO) \$(PERL) -Ilib -I"$local_lib_path" -I"$lib_path" -Iinc -MModule::Install::RTx::Runtime -e"RTxDatabase(qw(acl \$(NAME) \$(VERSION)))"
.
        $initdb .= <<"." if $has_etc{initialdata};
\t\$(NOECHO) \$(PERL) -Ilib -I"$local_lib_path" -I"$lib_path" -Iinc -MModule::Install::RTx::Runtime -e"RTxDatabase(qw(insert \$(NAME) \$(VERSION)))"
.
        $self->postamble("initdb ::\n$initdb\n");
        $self->postamble("initialize-database ::\n$initdb\n");
        if ($has_etc{upgrade}) {
            print "To upgrade from a previous version of this extension, use 'make upgrade-database'\n";
            my $upgradedb = qq|\t\$(NOECHO) \$(PERL) -Ilib -I"$local_lib_path" -I"$lib_path" -Iinc -MModule::Install::RTx::Runtime -e"RTxDatabase(qw(upgrade \$(NAME) \$(VERSION)))"\n|;
            $self->postamble("upgrade-database ::\n$upgradedb\n");
            $self->postamble("upgradedb ::\n$upgradedb\n");
        }
    }

}

sub requires_rt {
    my ($self,$version) = @_;

    _load_rt_handle();

    if ($self->is_admin) {
        $self->add_metadata("x_requires_rt", $version);
        my @sorted = sort RT::Handle::cmp_version $version,'4.0.0';
        $self->perl_version('5.008003') if $sorted[0] eq '4.0.0'
            and (not $self->perl_version or '5.008003' > $self->perl_version);
        @sorted = sort RT::Handle::cmp_version $version,'4.2.0';
        $self->perl_version('5.010001') if $sorted[0] eq '4.2.0'
            and (not $self->perl_version or '5.010001' > $self->perl_version);
    }

    # if we're exactly the same version as what we want, silently return
    return if ($version eq $RT::VERSION);

    my @sorted = sort RT::Handle::cmp_version $version,$RT::VERSION;

    if ($sorted[-1] eq $version) {
        die <<"EOT";

**** Error: This extension requires RT $version. Your installed version
            of RT ($RT::VERSION) is too old.

EOT
    }
}

sub requires_rt_plugin {
    my $self = shift;
    my ( $plugin ) = @_;

    if ($self->is_admin) {
        my $plugins = $self->Meta->{values}{"x_requires_rt_plugins"} || [];
        push @{$plugins}, $plugin;
        $self->add_metadata("x_requires_rt_plugins", $plugins);
    }

    my $path = $plugin;
    $path =~ s{\:\:}{-}g;
    $path = "$RT::LocalPluginPath/$path/lib";
    if ( -e $path ) {
        unshift @INC, $path;
    } else {
        my $name = $self->name;
        warn <<"EOT";

**** Warning: $name requires that the $plugin plugin be installed and
              enabled; it does not appear to be installed.

EOT
    }
    $self->requires(@_);
}

sub rt_too_new {
    my ($self,$version,$msg) = @_;
    my $name = $self->name;
    $msg ||= <<EOT;

**** Error: Your installed version of RT (%s) is too new; this extension
            only works with versions older than %s.

EOT
    $self->add_metadata("x_rt_too_new", $version) if $self->is_admin;

    _load_rt_handle();
    my @sorted = sort RT::Handle::cmp_version $version,$RT::VERSION;

    if ($sorted[0] eq $version) {
        die sprintf($msg,$RT::VERSION,$version);
    }
}

# RT::Handle runs FinalizeDatabaseType which calls RT->Config->Get
# On 3.8, this dies.  On 4.0/4.2 ->Config transparently runs LoadConfig.
# LoadConfig requires being able to read RT_SiteConfig.pm (root) so we'd
# like to avoid pushing that on users.
# Fake up just enough Config to let FinalizeDatabaseType finish, and
# anyone later calling LoadConfig will overwrite our shenanigans.
sub _load_rt_handle {
    unless ($RT::Config) {
        require RT::Config;
        $RT::Config = RT::Config->new;
        RT->Config->Set('DatabaseType','mysql');
    }
    require RT::Handle;
}

1;

__END__

=head1 NAME

Module::Install::RTx - RT extension installer

=head1 SYNOPSIS

In the F<Makefile.PL> of the C<RT-Extension-Example> module:

    use inc::Module::Install;
    RTx 'RT-Extension-Example';

    requires_rt '4.2.0';
    rt_too_new  '4.4.0';

    WriteAll();

=head1 DESCRIPTION

This B<Module::Install> extension implements several functions for
installing RT extensions:

=head2 RTx 'I<extension name>'

This function arranges for the following directories to be installed, if
they exist (assuming C<RTx('RT-Extension-Example')>):

    ./bin    => $RT::LocalPluginPath/RT-Extension-Example/bin
    ./etc    => $RT::LocalPluginPath/RT-Extension-Example/etc
    ./html   => $RT::LocalPluginPath/RT-Extension-Example/html
    ./lib    => $RT::LocalPluginPath/RT-Extension-Example/lib
    ./po     => $RT::LocalPluginPath/RT-Extension-Example/po
    ./sbin   => $RT::LocalPluginPath/RT-Extension-Example/sbin
    ./static => $RT::LocalPluginPath/RT-Extension-Example/static
    ./var    => $RT::LocalPluginPath/RT-Extension-Example/var

Accepts an optional argument hashref after the extension name with two possible keys

=head2 deprecated_rt

    If set to a true value, skips the enforced RT-4.0.0 minimum version check

    You should set a perl_version if using this option and requires_rt(), because requires_rt
    only handles figuring our what perl you need if you're on RT 4.0.0 or higher.

=head2 no_readme_generation

    If set to a true value, will not call readme_from on the extension's primary perl module.

=head2 requires_rt I<version>

Takes one argument, a valid RT version. If an attempt is made to install
on an RT than that version, it will die before Makefile creation.

=head2 requires_rt_plugin I<RT::Extension::Example> [, I<version>]

Ensures that the given RT extension (and optional version) is installed
in the target RT instance; C<requires> cannot be used because RT
extensions are not in @INC.

=head2 rt_too_new I<version> [, I<message>]

Takes one argument, a valid RT version, and prevents this module from
being installed on any version of RT equal to or newer than that.
Useful if a particular release of an extension only works on 4.0.x but
not 4.2.x.

Takes an optional second argument which allows you to specify a custom
error message. This message is passed to sprintf with two string
arguments, the current RT version and the version you specify.

=head2 remove_files

If set to a true value, during installation any files listed in a
C<remove_files> file will be removed from the destination directories.
This feature is for removing files that have been deleted or moved in
the current version and leaving the old version in place when upgrading
can cause problems.

RTx looks for a C<remove_files> file in the etc/upgrade directory of
the distribution. The format of the C<remove_files> file is as
follows:

    package Module::Install::RTx;

    @remove_files = qw(
    html/dir/file_to_remove
    );
    1;

The file locations are relative to the distribution. The destination
directory prefix is added automatically.

=head1 CAVEATS

=over 4

=item *

Us the full name when calling RTx method in Makefile.PL; while
C<RTx('Foo')> was once supported, it is no longer.

=back

=head1 ENVIRONMENT

=over 4

=item RTHOME

Path to the RT installation that contains a valid F<lib/RT.pm>.

=back

=head1 EXAMPLES

To install an extension which makes use of this installer:

    perl Makefile.PL RTHOME=/opt/rt4

This will install all subdirs into the $RT::LocalPluginPath dir
as configured in RT::Generated.

To install an extension into the (vendor) plugin path:

    perl Makefile.PL RTHOME=/opt/rt4 INSTALLDIRS=vendor

This will install all subdirs into the $RT::PluginPath which is specifically
meant for plugins that are installed through other packaging utils like
APT or RPM.

=head1 SEE ALSO

L<Module::Install>

L<http://www.bestpractical.com/rt/>

=head1 AUTHORS

Best Practical Solutions

(Originally) Audrey Tang <cpan@audreyt.org>

=head1 COPYRIGHT

Copyright 2003, 2004, 2007 by Audrey Tang E<lt>cpan@audreyt.orgE<gt>.
Copyright 2008-2014 Best Practical Solutions

This software is released under the MIT license cited below.

=head2 The "MIT" License

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.

=cut
